clang/lib/Sema/SemaTemplateInstantiateDecl.cpp

```diff
@@ -2572,7 +2572,7 @@ Decl *TemplateDeclInstantiator::VisitFunctionDecl(
   // Friend function defined withing class template may stop being function
   // definition during AST merges from different modules, in this case decl
   // with function body should be used for instantiation.
-  if (isFriend) {
+  if (isFriend && D->hasOwningModule()) {
```

Comment:
Clang AST is open to modifications.

Can't you add a new bit to `Redeclarable` to keep track of this?

Reply:
There is a problem with adding more flags to `FunctionDecl`, `FunctionDeclBitfields` is full and I cannot add there anything. `Redeclarable` doesn't have any flags as the moment and we need this extra flag only for FunctionDecls at the moment. So new version of the code works but most probably it cannot be committed as is. I also added another test case for the same example but inside module. My previous version was not able to compile it correctly.

Reply:
We can do this in ASTReader (ExternalASTSource). We can add a new interface to ExternalASTSource and calling it. It will be helpful since the call to `wasThisDeclarationADefinition()` should be cold.

Reply:
@ChuanqiXu9 sorry I'm not sure I fully understand your idea about ExternalASTSource. Are you suggesting to extend `ASTReader::DefinitionSource` to also keep this information there? Now `ASTReader::DefinitionSource` maps to `bool` but I can change value to a struct with two flags (previous DefinitionSource and ThisDeclarationWasADefinition), I think we don't need to store it in PCM file. Does it sound like the right approach?

Reply:
Oh yes, I took the last free bit in FunctionDeclBits a couple of months ago :)

This is a tall order, but it may be possible to make room in there. For example:

```C++
    LLVM_PREFERRED_TYPE(bool)
    uint64_t IsDefaulted : 1;
    LLVM_PREFERRED_TYPE(bool)
    uint64_t IsExplicitlyDefaulted : 1;
 ```
 
 It looks like `IsExplicitlyDefaulted` only makes sense if `IsDefaulted` is true, and maybe your new stuff only makes sense if `IsDefaulted` is false, so you can
probably share the `IsExplicitlyDefaulted` bit.

if that doesn't work, there are some other potential candidates.

For example, some bit fields only make sense if the function has a body, like `UsesSEHTry`, and these bits would probably be free for you as well.

Reply:
we can add bits by shrinking 

NumCtorInitializers.


however, in the present case it doesn't seem justified. these bits are precious.

Reply:
I would prefer ExternalASTSource approach because it won't add any additional memory overhead and we already have hashtable ASTReader::DefinitionSource. Access to the filed should be super cold so it won't slowdown things. I will update this PR soon.